skfda/misc/metrics/_basis_distance.py

```diff
@@ -0,0 +1,209 @@
+        \vdots & \vdots & \ddots & \vdots \\
+        \langle \phi_K, \phi_1 \rangle & \langle \phi_K, \phi_2 \rangle & \cdots & \langle \phi_K, \phi_K \rangle
+        \end{pmatrix}.
+
+    This class takes two `FDataBasis` objects with the same basis and computes 
+    the functional distance between them.
+
+    Attributes:
+        fd1 (FDataBasis): First functional data object.
+        fd2 (FDataBasis): Second functional data object.
+        weights (ndarray): Weighting function \( \nu_k \).
+        gram_matrix (ndarray): Gram matrix \( M \) of the basis functions.
+
+    Methods:
+        compute_distance(): Computes the functional distance \( d_{\nu}(X_1, X_2) \).
+
+    Example:
+
+        >>> from skfda.representation.basis import FDataBasis, BSplineBasis
+        >>> import numpy as np
+        >>> basis = BSplineBasis(n_basis=5)
+        >>> fd1 = FDataBasis(basis, np.array([1, 2, 3, 4, 5]))
+        >>> fd2 = FDataBasis(basis, np.array([2, 3, 1, 5, 4]))
+        >>> weights = np.ones(5)
+        >>> dist = BasisBasedDistance(fd1, fd2, weights)
+        >>> dist.compute_distance()
+        3.1622776601683795
+    """
+
+    def __init__(self, weights: Union[NDArrayFloat, None] = None) -> None:
+        self.weights = weights
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}(weights={self.weights})"
+
+    def __call__(self, fd1: FDataBasis, fd2: FDataBasis) -> NDArrayFloat:
+
+        if fd1.basis != fd2.basis:
+            raise ValueError("Both functional data objects must have the same basis.")
+
+        c1 = fd1.coefficients  # shape (n_samples_1, n_basis)
+        c2 = fd2.coefficients  # shape (n_samples_2, n_basis)
+
+        M = fd1.basis.gram_matrix()  # shape (n_basis, n_basis)
+
+        # Prepare weights
+        n_basis = c1.shape[1]
+        if self.weights is None:
+            weights = np.ones(n_basis)
+        else:
+            weights = np.zeros(n_basis)
+            weights[: self.weights.shape[0]] = self.weights
+
+        # If c2 has one sample and c1 has many, broadcast c2
+        if c1.shape[0] != c2.shape[0]:
+            if c2.shape[0] == 1:
+                c2 = np.repeat(c2, c1.shape[0], axis=0)
+            else:
+                raise ValueError("fd1 and fd2 must have the same number of samples or fd2 must have one sample.")
+
+        # Compute (c1 - c2) @ M
+        delta = np.dot((c1 - c2), M)  # shape (n_samples, n_basis)
+
+        # Weighted squared norm per sample
+        squared_dists = np.sum(weights * delta**2, axis=1)  # shape (n_samples,)
+
+        return np.sqrt(squared_dists)
+
+
+def basis_based_distance(
+    fd1: FDataBasis,
+    fd2: FDataBasis,
+    weights: Optional[NDArrayFloat] = None,
+) -> NDArrayFloat:
+    r"""
+    Computes the distance between two functional data objects represented in the same basis.
+
+    Args:
+        fd1 (FDataBasis): First functional data object.
```

Comment:
Remove the types from the docstring, Sphinx is able to add them from the type hints.